src/base/worker_pool_test.cc

```diff
@@ -0,0 +1,35 @@
+#include <base/thread.h>
+#include <base/worker_pool.h>
+
+#include <third_party/gtest/exported/include/gtest/gtest.h>
+
+namespace dist_clang {
+namespace base {
+
+TEST(WorkerPoolTest, InstantExitOnShutdown) {
+  Mutex mutex;
+  std::condition_variable condition;
+  bool ready = false;
+  bool done = false;
+  UniquePtr<WorkerPool> pool(new WorkerPool(true));
+
+  UniqueLock lock(mutex);
+
+  pool->AddWorker("TestWorker"_l, [&] (const WorkerPool& pool) {
+    UniqueLock lock(mutex);
+    condition.wait(lock, [&ready] { return ready; });
+    EXPECT_TRUE(pool.WaitUntilShutdown(std::chrono::seconds(5)));
+    done = true;
+    condition.notify_one();
+  });
+
+  Thread thread("Test"_l, [&] { pool.reset(); });
+  ready = true;
+  condition.wait_for(lock, std::chrono::seconds(1), [&done] { return done; });
```

Comment:
In this test this durations are expected as we need to check that worker exits early despite trying to wait for a long time. I've added another test below to check the opposite situation.